File: DevEnvironment/Base/src/main/java/base/core/AbstractPathBuilder.java
package base.core;

import base.interfaces.IPathBuilder;
import java.util.ArrayList;
import java.util.List;

/**
 * The AbstractPathBuilder base class implements an abstract builder of a path.
 */
public abstract class AbstractPathBuilder implements IPathBuilder {
    private final String rootDirectory;
    private final String separator;
    private final List<String> components = new ArrayList<>();
    private String extension;

    /**
     * The AbstractPathBuilder constructor.
     */
    protected AbstractPathBuilder(String rootDirectory, String separator) {
        Conditions.validateNotNull(
            separator,
            "The separator of a path.");

        this.rootDirectory = rootDirectory;
        this.separator = separator;
    }

    /**
     * Gets a root directory of a path.
     */
    @Override
    public String getRootDirectory() {
        return this.rootDirectory;
    }

    /**
     * Gets a separator of a path.
     */
    @Override
    public String getSeparator() {
        return this.separator;
    }

    /**
     * Adds a new component to a path.
     */
    @Override
    public IPathBuilder addComponent(String component) {
        Conditions.validateNotNull(
            component,
            "The component of a path.");

        int startIndex = 0;
        int endIndex = component.length() - 1;

        if (component.startsWith(this.separator)) {
            startIndex += this.separator.length();
        }

        if (component.endsWith(this.separator)) {
            endIndex -= this.separator.length();
        }

        int effectiveComponentLength = Indexes.size(startIndex, endIndex);

        if (effectiveComponentLength > 0) {
            String effectiveComponent = component.substring(startIndex, endIndex + 1);
            this.components.add(effectiveComponent);
        }

        return this;
    }

    /**
     * Sets an extension of a path.
     */
    @Override
    public IPathBuilder setExtension(String extension) {
        Conditions.validateNotNull(
            extension,
            "The extension of a path.");

        this.extension = extension;

        return this;
    }

    /**
     * Builds the resultant path.
     */
    @Override
    public String build() {
        StringBuilder result = new StringBuilder();

        if (!Strings.isNullOrEmpty(this.rootDirectory)) {
            result.append(this.rootDirectory);
        }

        boolean addDirectorySeparator = this.rootDirectory != null;
        for (String component : this.components) {
            if (addDirectorySeparator) {
                result.append(this.separator);
            }

            result.append(component);
            addDirectorySeparator = true;
        }

        if (!Strings.isNullOrEmpty(this.extension)) {
            result.append("." + this.extension);
        }

        return result.toString();
    }
}

File: DevEnvironmentCLion/Base/core/AbstractPathBuilder.cpp
#include "PreCompiled.h"
#include "AbstractPathBuilder.h"
#include "StringSearch.h"
#include "Indexes.h"

using namespace base;

/**
 * The AbstractPathBuilder Constructor.
 */
AbstractPathBuilder::AbstractPathBuilder(
    const std::string& rootDirectory,
    const std::string& separator) :
    m_rootDirectory(rootDirectory),
    m_separator(separator)
{
}

/**
 * The AbstractPathBuilder destructor.
 */
AbstractPathBuilder::~AbstractPathBuilder()
{
}

/**
 * Gets a root directory.
 */
const std::string& AbstractPathBuilder::getRootDirectory() const
{
    return m_rootDirectory;
}

/**
 * Gets a separator of a path.
 */
const std::string& AbstractPathBuilder::getSeparator() const
{
    return m_separator;
}

/**
 * Adds a new component to a path.
 */
IPathBuilder& AbstractPathBuilder::addComponent(const std::string& component)
{
    size_t startIndex = 0;
    size_t endIndex = component.size() - 1;

    if (StringSearch::startsWith(component, m_separator))
    {
        startIndex += m_separator.size();
    }

    if (StringSearch::endsWith(component, m_separator)) {
        endIndex -= m_separator.size();
    }

    size_t effectiveComponentLength = Indexes::size(startIndex, endIndex);

    if (effectiveComponentLength > 0) {
        std::string effectiveComponent = component.substr(startIndex, effectiveComponentLength);
        m_components.push_back(effectiveComponent);
    }

    return *this;
}

/**
 * Sets an extension of a path.
 */
IPathBuilder& AbstractPathBuilder::setExtension(const std::string& extension)
{
    m_extension = extension;
    return *this;
}

/**
 * Builds the resultant path.
 */
std::string AbstractPathBuilder::build()
{
    std::stringstream result;

    if (!m_rootDirectory.empty())
    {
        result << m_rootDirectory;
    }

    bool addSeparator = false;
    for (const std::string& component : m_components)
    {
        if (addSeparator)
        {
            result << m_separator;
        }

        result << component;
        addSeparator = true;
    }

    if (!m_extension.empty())
    {
        result << m_extension;
    }

    return result.str();
}
